### ISSUE TITLE
Sega Genesis Control Issue and Stuttering. 

### ISSUE THREAD

**Issue by QuietFrog2033** [jsgroth/jgenesis#135]
Hi, I am not sure if this is the right place to put this but, I'll give it a go.
I Start up jgenesis-0.8.1-windows-x86_64 and try to play Streets of Rage 2. 
When I play it I am using my Bluetooth 8-BitDo M30. When I lay out the buttons 
I am doing it correctly but, when I go to jump in the game it's C button. Punch is 
button B. When I go back to punch it thinks I am using both buttons which does a leg 
sweep.

I can't seem to get my punch button back again. It's like once I hit the jump button 
button C then I can't ever use the B button as punch. However, I can hold C button and hit 
the B button which does punch like normal but for whatever reason I can't use the B button 
alone anymore for punch. It only can be done while holding C and then hitting B. I don't know 
if this is a bug or I am doing something wrong. 

Last issue here, sometimes my Genesis games will stutter. I don't know if it's' me doing that with 
options but, it will stutter for some reason. It will take a little bit then clear up then I can play for a 
little bit and then it will slow and stutter again. I don't know what is going on as I am new to this 
emulator. I normally use Kega Fusion.

**Comment by jsgroth** [comment]
Hi,

Regarding the controls issue, I don't have an M30 and I'm not able to reproduce this with any of my controllers (including an 8BitDo SN30 Pro), but it definitely sounds like an issue in the emulator reading the controller inputs, like it's reading the C button inverted or something. It's really strange that it only seems to happen for one button - I'd expect an issue like this to happen for every button or none of them. I could try changing the gamepad implementation to poll the current SDL controller state instead of processing SDL controller events, but beyond that I'm not currently sure what I can do to fix this. I know the 8BitDo controllers have different modes (e.g. DInput / XInput / Switch) - have you tried switching the controller to a different mode to see if that helps at all?

Regarding the stuttering, is it slowing down (including audio stutter) or does it just look really choppy when it stutters? If it's actually slowing down then I would try disabling VSync in the video settings (I should probably have it disabled by default really). If it's choppy then you could try _enabling_ VSync, and if you're on Windows you could also try changing the wgpu backend to DirectX 12 - from my experience it seems to work a little better than the default Vulkan backend on Windows, although it probably depends on your GPU and drivers.

Improving the frame pacing when VSync is disabled is definitely something I want to fix - right now it seems to work well on some platforms/configurations but not so well on others. I haven't had much luck with the things I've tried so far though.

**Comment by QuietFrog2033** [comment]
Hello, hope you are having a great day. 
I'd like to say thank you for being so kind as to answer me and be so helpful. 
I sincerely appreciate that, and I wanted to let you know that before I give my answers. 
Ok, so the Controller I use is the 8Bit Do M30 it's a Sega Genesis style controller. In particular 
the 6-button style layout. This gamepad uses Bluetooth and is meant for Windows, Mac and 
Mobile phones both Android and Apple iPhone. On it is only the button to pair that's it for 
the controller. There are three buttons in the center but those are for mobile use. A Star, a 
minus symbol and a Heart symbol. Unfortunately, there isn't a switch to toggle on the controller.

As for the stuttering I will tell you what I tried. So, when booting up any game they all work so far 
as I tested. When I encountered the stuttering the first thing I did was to check the video settings. 
They were "out of the box" settings. Auto on most things and the rendering backends I have available
are: Vulkan, DirectX 12, and OpenGL. Now, when I was playing a game, and it would do the problem. 

I would quit and change the settings. I went with every single one of those options and it still had the issue. No matter 
what I did. As for Vsync, basically same outcome. It would still do it and while I did notice a smoother or less smooth
affect when toggling the Vsync it really didn't do much of anything to fix it. To save you reading time here, The rest of 
the options resulted in the same with just different styles of things going on in addition. Like interpolation etc. 

As for what it is doing. When I start a game like Super Street Fighter 2 The New challengers. The opening scene is great, 
smooth and looks good and the menus and selecting your character and all of that is great. When I get to a stage like this example. I will choose Ken, and it will go to Ryu's stage to face Ryu. I will throw my first fireball but after that if I don't get that move off first or really any move. The game will stutter and slow and the audio will go with it. It will sound like stuttering and the game will slow to a crawl. It's almost like in other Emulators where they compile shaders, and it slows
for a moment. 

But that's not what is going on here. We aren't compiling shaders for GameCube or PS3. We are talking about a 16-bit system. So, it will do this slow down and stutter sound in the game and in a minute, it clears up and you can play it again. Then it will do it again and you have to wait, and it can be frustrating as the next stage vs Chun-li you are trying to upper cut her while she does her Spin by the Feet or whatever move, and it does it and she beats you. 

So far this happens on any game that I have tested. They all have this issue of stuttering audio while slowing to a crawl and you have to wait for it to stop before you can play it again. I tried even starting the emulator with Windows compatibility mode to see if other windows vers would work. Nah, Same thing. I want to again say Thank you for taking the time to respond and for sitting through all of this text here. I wanted to be as honest as I could be and tell you what's happening in the best way I could. So, thank you for reading all of this. I know it's a lot. I hope anything I said can be helpful for you. I think what you are doing is a great and I love it very much.

**Comment by jsgroth** [comment]
What I meant by the M30 having different modes is that I think you can hold different buttons while powering it on to change what gamepad protocol it uses, if it's like 8BitDo's other wireless controllers. On the SN30 Pro for example you can hold X+Start for XInput, B+Start for DirectInput, Y+Start for Switch mode, and A+Start for MacOS mode. This _shouldn't_ make a huge difference (at least for XInput vs. DirectInput), but it's something you could try.

I might get an M30 myself to see if I can reproduce this - I already have a USB controller that works really well for Genesis/Saturn but it would be nice to have a wireless option too.

Thanks for the detail on the slowdown. It still sounds to me like either a sync issue or possibly a performance issue, though if it was performance-related then I would expect it to almost never run at full speed instead of sometimes slowing down and sometimes running at full speed.

Here are some other things you could try:
* Disable audio sync in the audio settings. This will cause audio crackling and will also make the game run way too fast if VSync is disabled, but if this fixes the slowdown, that means audio sync is the problem
* Disable rewind by changing the rewind buffer length to 0 in the hotkey settings. This shouldn't make a huge difference in terms of CPU (it only increases CPU usage by 1-2% last time I tested), but rewind does cause the emulator to use a lot of additional RAM

If audio sync ends up being the problem then I would try both decreasing the audio device queue size (32 or 64 should work fine) and increasing the audio sync threshold (20000 or so gives a larger buffer without increasing audio latency too much). I'm planning on changing the default values for these settings in the next release anyway as part of some tweaks to the audio playback code.

**Comment by jsgroth** [comment]
I got an M30 to test with and I found the source of the problem: SDL2 reads the M30's C button as an axis instead of a button. The axis input value goes straight from -32768 (min value) to +32767 (max value) when you press the button, and it instantaneously bounces back to -32768 when you release the button. I currently don't handle this correctly at all. The right shoulder button on this controller does the same thing (and has the same problem).

Fortunately, now that I can test this, it shouldn't be too hard to fix.

**Comment by jsgroth** [comment]
Here's a dev build that I believe fixes this bug with the M30 controller:
https://github.com/jsgroth/jgenesis/actions/runs/11925491297

I tested with a few different games/systems and it seems to work as expected now. The C button should map as "Axis 5 +".

I'd also be interested to know if the stuttering issues are any better with this build (or v0.8.2) - I've made a number of changes to video/audio sync since v0.8.1 that might help with that.

**Comment by jsgroth** [comment]
What are your PC specs, or at least OS and CPU?

I'm still not sure exactly what is causing this, but I just tested on the slowest computer I have that still works (i7-8650U laptop running Windows 10) and noticed some weird performance issues that might be related. With the laptop running on battery (lower CPU clock), the emulator pretty frequently drops down to 55-59 FPS, even though it runs at 95-100 FPS with the framerate uncapped. This isn't Genesis-specific, it also happens with the SNES core.

My best guess right now is that this is a bug in how the emulator implements sync on Windows, so I'm going to try improving that to see if that fixes this (or at least mitigates it).

**Comment by jsgroth** [comment]
I've pushed a change that tweaks how frame time sync works (https://github.com/jsgroth/jgenesis/actions/runs/12435651338) which mostly fixed the issue I mentioned in my previous comment. The emulator still hitches when running on battery and the laptop's power plan slows down the CPU too much (around 1 GHz seems to cause issues), but I don't think there's anything I can do about that short of general emulator performance improvements.

I tried doing the same thing for audio sync but that had a very negative impact on frame pacing, so I didn't commit those changes. It seems like SDL's audio playback code doesn't invoke the audio callback with consistent timing, which makes it difficult to precisely predict how long audio sync will need to wait for.

Re: perf improvements, I do have a few ideas for low-hanging improvements for Genesis specifically, so I'll work on those sooner rather than later. The VDP (video processor) in particular is implemented much less efficiently than it could be.

**Comment by SpelzR2** [comment]
> What are your PC specs, or at least OS and CPU?
Windows 11 & Solus (Linux)
AMD Radeon RX 7900 GRE
24 × AMD Ryzen 9 7900 12-Core Processor
32 GB RAM

I think the last version of jgenesis without the stuttering & slowdowns was 0.7.2. Since 0.8.0 I've these issues. I've also tried "Performance" profile on Win 11... didn't helped.

**Comment by jsgroth** [comment]
Okay, with that CPU it's almost definitely not a performance issue.

That's really helpful to know that you didn't have stuttering/slowdown issues in 0.7.2. There were a lot of changes between 0.7.2 and 0.8.0, but these are the ones that seem most suspect to me at first glance:
* https://github.com/jsgroth/jgenesis/commit/f16f3737: Changed backend cores to enqueue audio samples as they're generated instead of enqueueing once per frame
* https://github.com/jsgroth/jgenesis/commit/47bb6935: Changed audio sync to block at most once per frame rather than potentially being able to block every time audio samples are enqueued
  * I undid this change in 0.8.2, so it probably wasn't this (or at least not this alone)
* https://github.com/jsgroth/jgenesis/commit/f258769d and https://github.com/jsgroth/jgenesis/commit/c54b2dd8: Upgraded wgpu (graphics rendering library) from 0.18 to 0.20
  * Noting this because I've seen that some other projects experienced performance regressions after upgrading to wgpu 0.19
* https://github.com/jsgroth/jgenesis/commit/ddb25346: Disabled VSync by default (though this change wouldn't have done anything if you carried over a config file from a previous version)

Does the stuttering still occur in 0.8.2 or 0.8.3 if you enable frame time sync or VSync, and disable audio sync (all under Settings > Synchronization as of 0.8.2)? The audio-related changes reduced stuttering in my local testing but it might depend on audio device behavior. I have seen audio sync cause slowdown depending on the audio buffer sizes.

If this slowdown isn't related to audio sync then my next guess would be that something is getting stuck in the rendering layer, but that's going to be difficult to test. I can take a closer look at how behaviors changed between 0.7.2 and 0.8.0.

**Comment by SpelzR2** [comment]
I wanted to try out the different settings you've mentioned. So I booted Windows 11 on my PC and something magically happened... the stuttering was gone. It's strange since I remember I've tested 0.8.2 on Windows 11 in the past (since I have issue with jgenesis on Linux) and it stuttered back then. Only thing that happened in the meantime was the usual Windows updates and I've update my graphics card driver. I tried to be sure, so I've tested 0.8.1 / 0.8.2 & 0.8.3 and I can't reproduce the stuttering anymore. I've tried all different video API settings... didn't changed anything.

However since I've tested 0.8.3 on my somewhat older PC (with an Intel i7 870 2,93 GHz and an Nvidia Geforce GTX 960) shortly after its release I can confirm that at least on that machine I still encountered the stuttering / slow downs (regardless of Windows 10 or Debian 12). Unfortunately I only can test the settings later since that PC is not at my home anymore. I remember I've tried different settings like turning audio sync on/off or messing with the VSync settings with no success.

**Comment by jsgroth** [comment]
I'm glad you're not encountering the stuttering any more, but that's very strange that it just stopped happening for no apparent reason. I guess it's possible that one of the driver updates fixed it, but I'm struggling to think how the graphics driver could cause slowdown if VSync is disabled...unless maybe Windows or the driver was forcing VSync on and ignoring the emulator's VSync setting?

I guess it's also possible that some of the v0.8.0 audio changes could have caused lock contention between the main emulation thread and the SDL audio playback thread, but in my local testing I haven't seen that causing problems even when the audio hardware queue size is very very low (e.g. 8). I might increase the internal buffer size here a little just in case this was part of the issue - the audio latency impact of doing this is negligible (less than 1ms).

**Comment by SpelzR2** [comment]
I'm able to use my older PC for a couple of days. I've already wrote the specs of this machine (Intel i7 870 2,93 GHz and an Nvidia Geforce GTX 960 with Debian 12 KDE / Win 10). I've tested the latest build AppImage build of jgenesis (from 12/29) and the slowdowns /stuttering are still there. Neither changing the synchronization settings, nor disabling the rewind feature helps.

**Comment by jsgroth** [comment]
I'm mostly grasping at straws here, but I looked over the commit diff between v0.7.2 and v0.8.0 again, and I found another potential cause of significant performance regression: [subnormal floating-point numbers](https://en.wikipedia.org/wiki/Subnormal_number) in the audio resampling & filtering code. This code always runs, audio sync or not. This was not a new issue introduced in 0.8.0, but 0.8.0 did include some changes that could have made the impact much worse.

These are the two commits that stuck out (in addition to the ones I mentioned above): https://github.com/jsgroth/jgenesis/commit/eb57536fd05e70a3eb0aab04cf25b13ec2816928 and https://github.com/jsgroth/jgenesis/commit/7e5b7d605966e9f7c99e9701a304fab5bcddbf93, which respectively added more aggressive low-pass filters to audio output from the two Genesis sound chips. "More aggressive" in this context means more than twice as many filter coefficients, which means more than twice as many calculations in the audio output code that runs 48000 times per second. I adjusted these further in 0.8.3 but the sizes are close to the 0.8.0 filters.

I have no idea if this is what caused the particular stutter/slowdown you saw, but I tested a few games and there _are_ sometimes subnormal numbers in the audio filter calculations, which could definitely cause intermittent slowdown on an older CPU. Even on my fairly recent AMD CPU, if I force the non-AVX/FMA code path (which i7-870 would take because it doesn't support AVX), subnormals cause a _huge_ framerate drop if I'm running with no sync. From my cursory research, floating-point arithmetic instructions on subnormal numbers can take between 10 and _100_ times longer depending on the CPU, and the impact seems to be worst on older Intel CPUs.

Subnormals seem to most often appear after audio fade-outs, but the Streets of Rage games for example sometimes trigger them just with normal music playback which could cause slowdown during gameplay.

The apparently-intended way to avoid subnormals on x86 is to set the CPU's FTZ (flush to zero) and DAZ (denormals are zero) flags in MXCSR during the filter calculations, but thanks to LLVM assumptions it appears to be impossible to safely do this in current Rust without rewriting the entire filter routine in x86 assembly (see https://github.com/rust-lang/unsafe-code-guidelines/issues/471 and related), so instead I'm going to use the trick from this blog post to eliminate the subnormals:
https://www.earlevel.com/main/2019/04/19/floating-point-denormals/

Basically, alternate between adding and subtracting a very small value (e.g. 1e-15) to each generated sample before applying any filters, which prevents any of the filters from getting "stuck" at a subnormal value (or a value that's so small that it causes subnormals to appear in downstream calculations).

Again, not sure this is exactly what is causing your issue, but these subnormal numbers in the audio code are definitely a potential cause of intermittent awful performance.

**Comment by SpelzR2** [comment]
As of version 0.9.0 the sound issue / stuttering seems to be gone. I've tested some games (it was most prominent in Jurassic Park) and didn't encounter any problems on my older PC.

**Comment by jsgroth** [comment]
That's good to know! I bet it was the audio resampling changes, since I don't think I made any other changes between the new year and 0.9.0 that would have noticeably affected performance.

I'm not sure if the stuttering issue you encountered was exactly the same problem that was originally reported, but I'm going to go ahead and close this issue given that I fixed the M30 controller bug in 0.8.3 and the stuttering seems to be fixed as of 0.9.0. Feel free to comment or open another issue if the originally reported problem is not actually fixed as of 0.9.0.
